caffe2/opt/converter.cc

```diff
@@ -420,12 +442,88 @@ caffe2::NetDef convertToCaffe2Proto(repr::NNModule &m) {
   return convertToCaffe2Proto(m, predictNet);
 }
 
+bool shouldBeInplace(repr::NNGraph::NodeRef node) {
+  CAFFE_ENFORCE(repr::nn::is<repr::NeuralNetOperator>(node));
+  auto op = repr::nn::get<repr::NeuralNetOperator>(node);
+  using NNKind = repr::NeuralNetOperator::NNKind;
+  switch (op->getKind()) {
+    case NNKind::Relu:
+      return true;
+    default:
+      return false;
+  }
+}
+
+void addVersions(repr::NNModule* nn) {
+  std::unordered_map<std::string, uint64_t> versionMap;
+  for (auto& node : nn->dataFlow.getMutableNodes()) {
+    if (repr::nn::is<repr::NeuralNetData>(node)) {
+      auto d = repr::nn::get<repr::NeuralNetData>(node);
+      d->setVersion(versionMap[d->getName()]++);
+    }
+  }
+  for (auto& node_pair :
+       repr::nn::dataIterator<repr::NeuralNetOperator>(nn->dataFlow)) {
+    repr::NNGraph::NodeRef node;
+    repr::NeuralNetOperator* nnOp;
+    std::tie(nnOp, node) = node_pair;
+    // If we have an in-place node, all inputs/outputs should be versioned
+    // accordingly.
+    if (shouldBeInplace(node)) {
+      for (auto& input : repr::nn::getInputs(node)) {
+        for (auto& output : repr::nn::getInputs(node)) {
+          auto input_tensor = repr::nn::get<repr::NeuralNetData>(input);
+          auto output_tensor = repr::nn::get<repr::NeuralNetData>(output);
+          if (input_tensor->getName() == output_tensor->getName()) {
+            output_tensor->setVersion(input_tensor->getVersion());
```

Comment:
If output version equals to input version, we are not doing SSA in fact?

Reply:
yea, the idea of SSA is well preserved in nomnigraph (by having unique tensor node), but in caffe2 that model breaks for "in-place" operators which are annotated by an input output naming convention.  this is only necessary for caffe2

Reply:
If your the output of your in place op is also external output, how do we handle the logic?

Reply:
Yeah, in-place transform can be done "forward" and "backward". You need to try both. This is what Jessee and I have been doing for in-placing / removing ops such as StopGradient.